doc/bsetroot.1
.\"
.\" Man page for Bsetroot
.\"
.\" Copyright (c) 2000 by Wilbert Berendsen <wbsoft@xs4all.nl>
.\"
.TH bsetroot 1 "June 16th, 2000" "0.60.3"
.SH NAME
bsetroot \- blackbox utility to change root window appearance
.SH SYNOPSIS
.BR bsetroot " \-help"
.br
.B bsetroot
.RI "[ \-display " display " ] \-solid " color
.br
.B bsetroot
.RI "[ \-display " display " ] \-mod " "x y" " \-fg " color " \-bg " color
.br
.B bsetroot
.RI "[ \-display " display " ] \-gradient " texture " \-from " color " \-to " color
.SH DESCRIPTION
Bsetroot is a utility that can control the appearance of the root window in
three ways: Either give it a solid color, or write a two color modula pattern
to it, or render a gradient texture, based on two different colors.
.PP
Bsetroot resembles
.IR xsetroot (1)
in this functionality but it supports multiple screen displays, and gradient
textures the same way as Blackbox does.
It doesn't handle cursors etc.
Bsetroot is part of the Blackbox package.
.SH OPTIONS
Bsetroot operates in three ways, you must choose one of the first 3 options:
.TP
.BI \-solid " color"
Sets the root window to specified color.
.TP
.BI \-mod " x y"
Creates a modula pattern. You must specify
.BR \-bg " and " \-fg
colors.
.TP
.BI \-gradient " texturestring"
Renders the specified texture string to the root window.
For possible texture strings, please refer to
.IR blackbox (1).
You must also specify both a
.BR \-from " and a " \-to
color.
.TP
.BI \-display " display"
Tells Bsetroot to connect to the specified display.
.TP
.BI "\-bg, \-background " color
Background color.
Needed for
.B \-mod
patterns.
.TP
.BI "\-fg, \-foreground " color
Foreground color.
Needed for
.B \-mod
patterns.
.TP
.BI \-from " color"
Start color for rendering textures.
Needed for
.B \-gradient
operation mode.
.TP
.BI \-to " color"
Ending color for rendering textures.
Needed for
.B \-gradient
operation mode.
.TP
.B \-help
Prints version info and short help text.
.SH AUTHOR
Bsetroot was written and maintained by Brad Hughes
.nh \" hyphenation off
(blackbox@alug.org)
.hy \" on again
and Jeff Raven
.nh
(jraven@psu.edu).
.hy
.SH SEE ALSO
.IR blackbox (1)


NEWS
-- ChangeLog for Blackbox 0.70.x

Changes from 0.70.1 to blackbox-0.70.2
-----------------------------------------------

  - minor corrections to _NET_SUPPORTED and COMPLIANCE
  - prepare to support _NET_STARTUP_ID and friends
  - support _NET_WM_USER_TIME/_NET_WM_USER_TIME_WINDOW
  - added support for _NET_FRAME_EXTENTS
  - fix _NET_DESKTOP_NAMES bug
  - correct typo
  - tweak compliance
  - more EWMH support: all types, states and actions
  - add support for additional _NET_WM_WINDOW_TYPEs
  - added WM_Sn MANAGER selection compliance
  - prepare for _NET_SHOWING_DESKTOP support
  - add missing EWMH atoms
  - add support for _NET_WM_STRUT_PARTIAL
  - added support for _NET_WM_WINDOW_TYPE_NOTIFICATION
  - marked more EWMH support
  - typo in patch
  - correct compiler warnings
  - automake corrections
  - autoconf corrections
  - applied nls-codesets.patch from pld-linux
  - fix runaway workspace creation ewmh.patch
  - applied patch from github.com/tkoskine/blackboxwm
  - applied bbdock.patch from archlinux
  - applied bsetbg-feh.patch from archlinux
  - applied textpropertytostring-unconditional.patch from debian

  Following changes from blackboxwm CVS:

  - apply patch from debian to compile with gcc 4.3.0
  - fix typo
  - apply patch from 1696203, which enables window snapping for
    windows on all workspaces
  - #include <stdlib.h>, as pointed out in bug 1709671
  - XFT is a compile time option, show it in -v output
  - s,Bitstream Vera Sans,Sans Serif,g
  - run on kubuntu
  - avoid dangling pointers by setting them to zero after deleting
  - silence warnings from gcc 4.0.2
  - compile with gcc 2.9x
  - get rid of the GC/Xft cache... it's a constant source of
    trouble and doesn't really give us a whole lot otherwise

Changes from 0.70.0 to 0.70.1:
------------------------------------------------------------------------------

  - fix compilation on OpenBSD and Mac OS X
  - documentation updates to data/README.menu
  - fix bt::Timer to make sure it still works after daylight-savings
    changes
  - fix horizontal text alignment by adding a small indent to all text
  - fix sunken gradient rendering: the colors are no longer inverted,
    and the correct bevel is drawn
  - add bt::XDG::BaseDir, which is an implementation of the
    freedesktop.org basedir-spec version 0.6
  - make bt::Pen able to recover from cache faults
  - silence messages from bt::PixmapCache about the cache maximum
    being exceeded
  - add bt::Rect::inside(), which ensures that one rectangle is inside
    the other
  - fix bt::Resource to do nothing if filename arguments are empty
  - fix bt::Resource::merge() to stop overriding existing entries (it
    should only add new ones)
  - add bt::Texture::textureResource() overload that returns a default
    texture if the resource cannot be found
  - add bt::dirname(), which returns the directory part of a file path
  - add bt::mkdirhier(), which is equivalent to
    'mkdir -m mode -p path'
  - add bt::tolower(), which returns a string with all characters
    lowercase
  - add the "Center Placement" window placement policy, which places
    all new windows in the center of the screen
  - add "Shade Window with Mouse Wheel" and "Toolbar Actions with
    Mouse Wheel" configuration options
  - add window-to-window snapping support. This feature is off by
    default, you need to set session.windowSnapThreshold to a positive
    number in your ~/.blackboxrc.
  - add support for piped menus. The output of a command can be used
    to create the blackbox menu. This is done by prepending the pipe
    character, |, to session.menuFile in ~/.blackboxrc or to [include]
    filenames.
  - [nop] menu entries are now drawn disabled
  - fix "Focus Last Window on Workspace" behavior to stop preferring
    windows that occupy all workspaces
  - window placement takes windows with Desktop and Dialog types into
    consideration and does the appropriate thing (instead of placing
    them normally)
  - remove 1 pixel gaps between windows when doing smart placement
  - allow users to override the style's rootCommand in their
    ~/.blackboxrc
  - fix default colors used by blackbox when no style is found
  - allow the Toolbar to be raised and lowered with
    Alt+LeftMouseButton and Alt+MiddleMouseButton, respectively
  - fix maximized+shaded windows from disappearing when restarting
    blackbox
  - each window's windowmenu can be accessed now with
    Mod4+RightMouseButton
  - fix fullscreen support to ignore aspect ratios set by applications
    (specifically, this should fix movie players that are currently
    displaying 16:9 video on a 4:3 screen)
  - fix support for _NET_MOVERESIZE_WINDOW, normally sent from pagers
  - fix focus problems caused by applications that try to fake
    fullscreen by setting focus to an override-redirect window
    (e.g. rdekstop)

Changes from 0.65.x to 0.70.0:
------------------------------------------------------------------------------

This is a major update resulting from a long development cycle... the
list below is a general list and does not cover all changes in detail.

  - introduced a utility library (libbt) for aiding in writing
    small blackbox oriented applications
  - all menus are now proper popup menus, with keyboard navigation
  - complete EWMH support
  - Xft support has been added to support anti-aliased fonts
  - support for texture borders has been added... any texture can now
    have a border of arbitrary width and color
  - ~/.blackboxrc is no longer written on exit or restart.  Instead it
    is written whenever a change is made.  This means the rc can be
    edited while Blackbox is running (finally).
  - support for always-on-top and always-on-bottom has been added
  - support for fullscreen windows has been added
  - support for all EWMH window types has been added (e.g. desktop
    windows, dock windows, dialogs, etc.)
  - support for window groups has been added

util/bsetroot.cc
// -*- mode: C++; indent-tabs-mode: nil; c-basic-offset: 2; -*-
// bsetroot - a background setting utility
// Copyright (c) 2001 - 2005 Sean 'Shaleh' Perry <shaleh at debian.org>
// Copyright (c) 1997 - 2000, 2002 - 2005
//         Bradley T Hughes <bhughes at trolltech.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#include "gettext.h"
#include "bsetroot.hh"

#include <Pen.hh>
#include <Texture.hh>

#include <cstdlib>
#include <cctype>

#include <X11/Xatom.h>
#include <cstdio>
#include <cstdlib>

#include <cstring>

// ignore all X errors
static int x11_error(::Display *, XErrorEvent *)
{ return 0; }


bsetroot::bsetroot(int argc, char **argv, char *dpy_name,
                   bool multi_head): display(dpy_name, multi_head) {
  XSetErrorHandler(x11_error); // silently handle all errors

  bool mod = False, sol = False, grd = False;
  int mod_x = 0, mod_y = 0;

  for (int i = 1; i < argc; i++) {
    if (! strcmp("-help", argv[i])) {
      usage();
    } else if ((! strcmp("-fg", argv[i])) ||
               (! strcmp("-foreground", argv[i])) ||
               (! strcmp("-from", argv[i]))) {
      if ((++i) >= argc) usage(1);

      fore = argv[i];
    } else if ((! strcmp("-bg", argv[i])) ||
               (! strcmp("-background", argv[i])) ||
               (! strcmp("-to", argv[i]))) {
      if ((++i) >= argc) usage(1);

      back = argv[i];
    } else if (! strcmp("-solid", argv[i])) {
      if ((++i) >= argc) usage(1);

      fore = argv[i];
      sol = True;
    } else if (! strcmp("-mod", argv[i])) {
      if ((++i) >= argc) usage();

      mod_x = atoi(argv[i]);

      if ((++i) >= argc) usage();

      mod_y = atoi(argv[i]);

      if (mod_x < 1) mod_x = 1;
      if (mod_y < 1) mod_y = 1;

      mod = True;
    } else if (! strcmp("-gradient", argv[i])) {
      if ((++i) >= argc) usage();

      grad = argv[i];
      grd = True;
    } else if (! strcmp("-display", argv[i])) {
      // -display passed through tests ealier... we
      i++;
    } else {
      usage();
    }
  }

  if ((mod + sol + grd) != True) {
// TRANS Option error message.
    fprintf(stderr,
            gettext("bsetroot: error: must specify one of: -solid, -mod, -gradient\n"));

    usage(2);
  }

  // keep the server grabbed while rendering all screens
  XGrabServer(display.XDisplay());

  if (sol && ! fore.empty())
    solid();
  else if (mod && mod_x && mod_y && ! (fore.empty() || back.empty()))
    modula(mod_x, mod_y);
  else if (grd && ! (grad.empty() || fore.empty() || back.empty()))
    gradient();
  else
    usage();

  // ungrab the server and discard any events
  XUngrabServer(display.XDisplay());
  XSync(display.XDisplay(), True);
}


bsetroot::~bsetroot(void) {
  XSetCloseDownMode(display.XDisplay(), RetainPermanent);
  XKillClient(display.XDisplay(), AllTemporary);
}


// adapted from wmsetbg
void bsetroot::setPixmapProperty(int screen, Pixmap pixmap) {
  Atom rootpmap_id = XInternAtom(display.XDisplay(), "_XROOTPMAP_ID", False),
       esetroot_id = XInternAtom(display.XDisplay(), "ESETROOT_PMAP_ID", False);

  const bt::ScreenInfo &screen_info = display.screenInfo(screen);

  Atom type;
  int format;
  unsigned long length, after;
  unsigned char *data = 0;
  Pixmap xrootpmap = None;
  Pixmap esetrootpmap = None;

  // Clear out the old _XROOTPMAP_ID property
  XGetWindowProperty(display.XDisplay(), screen_info.rootWindow(),
		     rootpmap_id, 0L, 1L, True, AnyPropertyType,
                     &type, &format, &length, &after, &data);

  if (data && type == XA_PIXMAP && format == 32 && length == 1)
    xrootpmap = *(reinterpret_cast<Pixmap *>(data));

  if (data) XFree(data);

  // Clear out the old ESETROOT_PMAP_ID property
  XGetWindowProperty(display.XDisplay(), screen_info.rootWindow(),
                     esetroot_id, 0L, 1L, True, AnyPropertyType,
                     &type, &format, &length, &after, &data);

  if (data && type == XA_PIXMAP && format == 32 && length == 1)
    esetrootpmap = *(reinterpret_cast<Pixmap *>(data));

  if (data) XFree(data);

  // Destroy the old pixmaps
  if (xrootpmap)
    XKillClient(display.XDisplay(), xrootpmap);
  if (esetrootpmap && esetrootpmap != xrootpmap)
    XKillClient(display.XDisplay(), esetrootpmap);

  if (pixmap) {
    XChangeProperty(display.XDisplay(), screen_info.rootWindow(),
		    rootpmap_id, XA_PIXMAP, 32, PropModeReplace,
		    reinterpret_cast<unsigned char *>(&pixmap), 1);
    XChangeProperty(display.XDisplay(), screen_info.rootWindow(),
		    esetroot_id, XA_PIXMAP, 32, PropModeReplace,
		    reinterpret_cast<unsigned char *>(&pixmap), 1);
  }
}


unsigned long bsetroot::duplicateColor(unsigned int screen,
                                       const bt::Color &color) {
  /*
    When using a colormap that is not read-only, we need to
    reallocate the color we are using.  The application's color cache
    will be freed on exit, and another client can allocate a new
    color at the same pixel value, which will immediately change the
    color of the root window.

    this is not what we want, so we need to make sure that the pixel
    we are using is doubly allocated, so that it stays around with the
    pixmap.  It will be released when we use
    XKillClient(..., AllTemporary);
  */
  const bt::ScreenInfo &screen_info = display.screenInfo(screen);
  unsigned long pixel = color.pixel(screen);
  XColor xcolor;
  xcolor.pixel = pixel;
  XQueryColor(display.XDisplay(), screen_info.colormap(), &xcolor);
  if (! XAllocColor(display.XDisplay(), screen_info.colormap(),
                    &xcolor)) {
    fprintf(stderr, "warning: couldn't duplicate color %02x/%02x/%02x\n",
            color.red(), color.green(), color.blue());
  }
  return pixel;
}


void bsetroot::solid(void) {
  bt::Color c = bt::Color::namedColor(display, 0, fore);

  for (unsigned int screen = 0; screen < display.screenCount(); screen++) {
    const bt::ScreenInfo &screen_info = display.screenInfo(screen);
    unsigned long pixel = duplicateColor(screen, c);

    XSetWindowBackground(display.XDisplay(), screen_info.rootWindow(), pixel);
    XClearWindow(display.XDisplay(), screen_info.rootWindow());

    Pixmap pixmap =
      XCreatePixmap(display.XDisplay(), screen_info.rootWindow(),
                    8, 8, DefaultDepth(display.XDisplay(), screen));

    bt::Pen pen(screen, c);
    XFillRectangle(display.XDisplay(), pixmap, pen.gc(), 0, 0, 8, 8);

    setPixmapProperty(screen, pixmap);
  }
}


void bsetroot::modula(int x, int y) {
  char data[32];
  long pattern;

  unsigned int screen, i;

  bt::Color f = bt::Color::namedColor(display, 0, fore),
            b = bt::Color::namedColor(display, 0, back);

  for (pattern = 0, screen = 0; screen < display.screenCount(); screen++) {
    for (i = 0; i < 16; i++) {
      pattern <<= 1;
      if ((i % x) == 0)
        pattern |= 0x0001;
    }

    for (i = 0; i < 16; i++) {
      if ((i %  y) == 0) {
        data[(i * 2)] = static_cast<char>(0xff);
        data[(i * 2) + 1] = static_cast<char>(0xff);
      } else {
        data[(i * 2)] = pattern & 0xff;
        data[(i * 2) + 1] = (pattern >> 8) & 0xff;
      }
    }

    const bt::ScreenInfo &screen_info = display.screenInfo(screen);

    XGCValues gcv;
    gcv.foreground = duplicateColor(screen, f);
    gcv.background = duplicateColor(screen, b);

    GC gc = XCreateGC(display.XDisplay(), screen_info.rootWindow(),
                      GCForeground | GCBackground, &gcv);

    Pixmap pixmap = XCreatePixmap(display.XDisplay(),
				  screen_info.rootWindow(),
				  16, 16, screen_info.depth());

    Pixmap bitmap =
      XCreateBitmapFromData(display.XDisplay(), screen_info.rootWindow(),
                            data, 16, 16);
    XCopyPlane(display.XDisplay(), bitmap, pixmap, gc,
               0, 0, 16, 16, 0, 0, 1l);
    XFreeGC(display.XDisplay(), gc);
    XFreePixmap(display.XDisplay(), bitmap);

    XSetWindowBackgroundPixmap(display.XDisplay(),
                               screen_info.rootWindow(),
                               pixmap);
    XClearWindow(display.XDisplay(), screen_info.rootWindow());

    setPixmapProperty(screen, pixmap);
  }
}


void bsetroot::gradient(void) {
  /*
    we have to be sure that neither raised nor sunken is specified otherwise
    odd looking borders appear.  So we convert to lowercase then look for
    'raised' or 'sunken' in the description and erase them.  To be paranoid
    the search is done in a loop.
  */
  std::string descr = bt::tolower(grad);
  std::string::size_type pos;
  while ((pos = descr.find("raised")) != std::string::npos)
    descr.erase(pos, 6); // 6 is strlen raised
  while ((pos = descr.find("sunken")) != std::string::npos)
    descr.erase(pos, 6);

  // now add on 'flat' to prevent the bevels from being added
  descr += "flat";

  bt::Color f = bt::Color::namedColor(display, 0, fore);
  bt::Color b = bt::Color::namedColor(display, 0, back);

  bt::Texture texture;
  texture.setDescription(descr);
  texture.setColor1(f);
  texture.setColor2(b);

  for (unsigned int screen = 0; screen < display.screenCount(); screen++) {
    const bt::ScreenInfo &screen_info = display.screenInfo(screen);

    bt::Image image(screen_info.width(), screen_info.height());
    Pixmap pixmap = image.render(display, screen, texture);

    XSetWindowBackgroundPixmap(display.XDisplay(),
                               screen_info.rootWindow(),
                               pixmap);
    XClearWindow(display.XDisplay(), screen_info.rootWindow());

    setPixmapProperty(screen, pixmap);
  }
}


void bsetroot::usage(int exit_code) {
  fprintf(stderr,
          "bsetroot 3.1\n\n"
          "Copyright (c) 2001 - 2005 Sean 'Shaleh' Perry\n"
          "Copyright (c) 1997 - 2000, 2002 - 2005 Bradley T Hughes\n");
// TRANS Help message for the bsetroot program.
  fprintf(stderr, gettext(
          "  -display <string>        use display connection\n"
          "  -mod <x> <y>             modula pattern\n"
          "  -foreground, -fg <color> modula foreground color\n"
          "  -background, -bg <color> modula background color\n\n"
          "  -gradient <texture>      gradient texture\n"
          "  -from <color>            gradient start color\n"
          "  -to <color>              gradient end color\n\n"
          "  -solid <color>           solid color\n\n"
          "  -help                    print this help text and exit\n"));
  exit(exit_code);
}

int main(int argc, char **argv) {
  char *display_name = 0;
  bool multi_head = False;

#ifdef ENABLE_NLS
  bindtextdomain(PACKAGE, LOCDIR);
  textdomain(PACKAGE);
#endif
  for (int i = 1; i < argc; i++) {
    if (! strcmp(argv[i], "-display")) {
      // check for -display option

      if ((++i) >= argc) {
        fprintf(stderr, "error: '-display' requires an argument\n");

        ::exit(1);
      }

      display_name = argv[i];
    } else if (! strcmp(argv[i], "-multi")) {
      multi_head = True;
    }
  }

  bsetroot app(argc, argv, display_name, multi_head);
  return 0;
}
